crates/qasm3/src/build.rs

```diff
@@ -377,37 +372,45 @@ pub fn convert_asg(
         pygates: gate_constructors,
         module,
     };
+
+    let _ = state.map_gate_ids(py, ast_symbols);
+
     for statement in program.stmts().iter() {
         match statement {
             asg::Stmt::GateCall(call) => state.call_gate(py, ast_symbols, call)?,
             asg::Stmt::DeclareClassical(decl) => state.declare_classical(py, ast_symbols, decl)?,
             asg::Stmt::DeclareQuantum(decl) => state.declare_quantum(py, ast_symbols, decl)?,
-            asg::Stmt::GateDeclaration(decl) => state.define_gate(py, ast_symbols, decl)?,
+            // We ignore gate definitions because the only information we can currently use
+            // from them is extracted with `SymbolTable::gates` via `map_gate_ids`.
+            asg::Stmt::GateDefinition(_) => (),
```

Comment:
Doing this is fine, but leads me to check: does the backing parser correctly handle the order of definitions?  For example, will it correctly reject the program
```qasm
qubit q;
my_gate q;
gate my_gate q {}
```
because `my_gate` isn't defined at the point of use?

I'm guessing so - I'm just checking because this swap from a single-pass iteration in the Qiskit side to relying on the AST means that the responsibility for handling it moves away from us.

Reply:
Yes it is rejected:
```
Error: UndefGateError
   ╭─[/path/to/examples/qasm/ex1.qasm:2:1]
 2 │my_gate q;
   │   ╰──── Near this point
```
There were some other doubts I had, a bit vague, about this approach. I don't recall what they are. But if they are warranted, it should become clear later.